File: benchmarks/python/02_cohort_generation.py
"""Benchmark 02: Cohort Generation — omopy.connector.generate_concept_cohort_set()"""

import polars as pl
from helpers import Timer, connect_cdm, save_result, save_timing

from omopy.connector import generate_concept_cohort_set
from omopy.generics import Codelist

print("=== 02: Cohort Generation ===")
t = Timer()
cdm = connect_cdm()

codelist = Codelist({"coronary_artery": [317576]})
cdm = generate_concept_cohort_set(cdm, codelist, name="target_cohort")
cohort = cdm["target_cohort"]

df = cohort.collect()
counts = df.group_by("cohort_definition_id").agg(
    pl.len().alias("n_records"),
    pl.col("subject_id").n_unique().alias("n_subjects"),
)
save_result(counts, "02_cohort_generation")
save_timing("02_cohort_generation", t.elapsed())
print(f"Done in {t.elapsed():.2f} seconds")

File: benchmarks/r/02_cohort_generation.R
# Benchmark 02: Cohort Generation
# R equivalent: CDMConnector::generateConceptCohortSet()

source("benchmarks/r/00_helpers.R")
cat("=== 02: Cohort Generation ===\n")
t0 <- proc.time()

con <- dbConnect(duckdb(), dbdir = DB_PATH)
cdm <- cdmFromCon(con = con, cdmSchema = "main", writeSchema = "main", cdmName = "synthea_1k")

# Coronary arteriosclerosis (317576) — 1243 records in this DB
cdm <- generateConceptCohortSet(
  cdm = cdm,
  conceptSet = list(coronary_artery = 317576),
  name = "target_cohort"
)

counts <- cdm$target_cohort |>
  group_by(cohort_definition_id) |>
  summarise(n_records = n(), n_subjects = n_distinct(subject_id)) |>
  collect()

save_result(counts, "02_cohort_generation")
elapsed <- (proc.time() - t0)[["elapsed"]]
save_timing("02_cohort_generation", elapsed)
cdmDisconnect(cdm)
cat("Done in", elapsed, "seconds\n")